working/extension_structs/overview.md

```diff
@@ -0,0 +1,815 @@
+and provides a number of convenient methods.
+
+```dart
+void test() {
+  // Call the default constructor, using the default value for w
+  var c1 = Component(0, 0, 0);
+  // Call the default constructor, explicitly passing w
+  var c2 = Component(1, 1, 1, 1);
+
+  assert(c1 == c2); // Equality is defined to be structural
+  assert(c1.hashCode == c2.hashCode); // With correspondind hashCode
+
+  print(c1.debutToString()); // Prints "Component(0, 0, 0, 1)"
+
+  print(c1.toString()); // Prints "struct"
+
+  // Generated copy method
+  var c2 = c1.copyWith(x : 2);
+  assert(c2.x == 2);
+}
+```
+
+Structs may explicitly define constructors, static members, and normal class
+members (except fields), and may implement interfaces.
+
+```dart
+struct Component(int _x, int _y, int _z, int _w = 1)
+  implements Comparable<Component> {
+
+  double get x => _x/_w;
+  double get y => _x/_w;
+  double get z => _z/_w;
+
+  int compareTo(Component other) => throw "TODO: Unimplemented";
+
+  Component.zero() : _x = 0, _y = 0, _z = 0;
+}
+```
+
+With the addition of pattern matching and switches, structs will support closed
+families (algebraic datatypes).
+
+```dart
+abstract struct Operand;
+
+struct ConstantOperand(Value c) extends Operand;
+
+struct IdentifierOperand(Identifier i) extends Operand;
+
+Operand replaceIn(Operand oper, Identifier t, Value c) {
+  return switch (oper) {
+    case ConstantOperand(_) : oper;
+    case IdentifierOperand(var i) where i == t: ConstantOperand(c);
+    case IdentifierOperand(_) : oper;
+  }
+}
+```
+
+## Structs in more detail
+
+### Introduction form
+
+We add a new keyword "struct", which is used in place of "class".  The class
+name (plus generics if applicable) must be followed by a parenthesized list of
+field declarations (more on this below). There may optionally be an "extends"
+clause, and an "implements" list.  No mixins are permitted.  After the header,
+the usual brace delimited list of members may be provided subject to
+restrictions described below.  An empty set of members may be elided in favor of
+a semi-colon.  The simplest possible struct definitions then would look like:
+
+```dart
+struct Data();
+struct GenericData<T>();
+```
+
+#### Alternatives
+
+
+##### Re-use the class keyword
+
+An alternative is to continue to use "class" and to use some other piece of
+syntax to indicate that the object in question is not a general class.  For
+example, the parenthesized list might be enough:
+
+
+```dart
+class Data();
+class GenericData<T>();
+```
+
+Alternatively, an additional keyword could be used:
+
+```dart
+class Data wraps ();
+class GenericData<T> wraps ();
+```
+
+Other keywords could be considered.
+
+##### Choose a different leading keyword
+
+Instead of "struct", we could use "view", or some other choice:
+
+```dart
+view Data();
+view GenericData<T>();
+```
+
+##### Use a modifier on classes
+
+
+We could use "data class":
+
+```dart
+data class Data();
```

Comment:
One thing that draws me to this syntax is that it aligns with the vision that this is a "restrictive" feature on top of what already exists.

Now I can map that a `data class` is a `class` with the following restrictions:
* it can only extend from an empty `data class` that acts as a tagging interface
* it can implement interfaces and override methods like regular classes
* unless otherwise specified, the fields by default structurally identify instances of the data class
* it doesn't export a default interface, and hence cannot be "implemented" by other classes in the program

Similarly a `static data class` (using the proposed `static` modifier i mentioned earlier) can be explained as a `data class` that:
* only supports having a single field (a wrapped object)
* always implements a single interface: the type of the wrapped object (called the wrapped type)
* doesn't support virtual method dispatch, except for the methods in the wrapped type
* cannot shadow/override methods of the wrapped type, but may statically expose a subset of them.
* has its type available at compile-time only, but later gets erased and replaced by the wrapped type at runtime.

Reply:
Yeah, I could definitely get behind this story.

Reply:
> One thing that draws me to this syntax is that it aligns with the vision that this is a "restrictive" feature on top of what already exists.

I don't see much value in having a feature that users have to opt in to using and in return... it just prevents them from doing things. I think restrictive features makes sense when in return for fitting within some restrictions, you get other *addition* for free. If it's just "add this modifier and the compiler will yell at you more", I don't find that very compelling. I understand the general software engineering argument for restrictions making things easier to maintain, but I don't seem those really applying to data classes. A *user* of a data class shouldn't care whether it was implemented using a `data` modifier or by hand.

What I *do* find very compelling is, "Add this modifier and you'll get implementations of `==`, `hashCode`, and `copyWith()` for free." That's essentially what `data` does in Kotlin and `record` in Java and C#. Ideally, any time a user wants those methods with the behavior that the syntactic sugar would produce, then they should be able to use this modifier to get it. The fewer restrictions involved, the more often they'll be able to reach for this modifier and get nicer code.

So, in order to provide implementations of those methods, does the language really need to care if all the fields are final or if the class has an implicit interface? In other words, which of these restrictions are essential?

Reply:
I think you're misunderstanding the usage of "restrictive", which comes from me.  What I mean by that is not "I restrict you from doing things!", but rather "as much as possible, these things behave exactly like classes, which you are familiar with, except that there are some things you can't do anymore, in order that we may provide you with these benefits".  This is in contrast to some other proposals we've made in this area where "extends" or "implements" meant something fairly substantially different.

> If it's just "add this modifier and the compiler will yell at you more", I don't find that very compelling

I don't think there's anything that I'm doing in here where this is the motivation.

> The fewer restrictions involved, the more often they'll be able to reach for this modifier and get nicer code.

Agreed.  But also, the more benefits they get, they more often they'll reach for this modifier and get nicer code.  Essentially all of my choices are (at least intended to be) driven by benefits.

> So, in order to provide implementations of those methods, does the language really need to care if all the fields are final or if the class has an implicit interface? In other words, which of these restrictions are essential? (I mean, obviously, a mutable class with value-based == and hashCode is gonna behave weird if you mutate it. But if the user wants to do that because they know they'll never stuff it in a hash table, why not let them?)

It is very much a goal of this to provide/allow structural identity.  If you don't agree with that goal, then you won't see that as a benefit (but I do).  If you take that as a goal, the semantics get insane to a degree that I'm not comfortable with if you allow mutability.

For the implicit interface, there are at least three aspects that I see as the benefit: one is performance, another is that you get ADTs (sealed families) for free, another is that you get field promotion for free.  We could certainly make the ADT thing opt in.  I don't love that it requires yet more boilerplate, but I'm not intensely opposed.  For field promotion.... maybe we add something to interfaces that requires them to be "stable"?  But then there's more leakage.